### PR TITLE
feat(product): Apply transaction decorators to the services

### PR DESCRIPTION
**Pull request by adrien2p**
Use the InjectEntityManager to manage the transactions

### CODE REVIEW

**Comment by olivermrbl** [comment]
Can I get you to add what and why in the PR description 🙏

**Comment by adrien2p** [comment]
do not merge yet

**Comment by adrien2p** [comment]
@olivermrbl it should be good to re review :)

**Review by carlos-r-l-rodrigues — APPROVED** [review]
LGTM

**Review by olivermrbl — COMMENTED** [review]
wondering if we can somehow infer the second parameter of the `InjectEntityManage` or improve the way it's passed to make it less error-prone (error-prone -> using raw strings as we've discussed in other contexts)

**Comment by adrien2p** [comment]
> wondering if we can somehow infer the second parameter of the `InjectEntityManage` or improve the way it's passed to make it less error-prone (error-prone -> using raw strings as we've discussed in other contexts)

I don't see a way to infer it correctly as it can be whatever property the user want to assign the object to

File: packages/product/src/repositories/product-tag.ts [hunk 1]
```diff
@@ -91,32 +94,43 @@ export class ProductTagRepository extends AbstractBaseRepository<ProductTag> {
...
 
     return upsertedTags
   }
 
-  async delete(ids: string[], context: Context = {}): Promise<void> {
-    const manager = (context.transactionManager ??
-      this.manager_) as SqlEntityManager
-
-    await manager.nativeDelete(Product, { id: { $in: ids } }, {})
+  @InjectEntityManager(doNotForceTransaction, "")
```
Comment (olivermrbl) [inline packages/product/src/repositories/product-tag.ts]: I don't love the empty string here. Can we make the argument optional for these cases?

**Review by olivermrbl — COMMENTED** [review]
Do we need to type-cast all repositories going forward, because we use the DAL types on class members? It's fine with me, just messes up the code a bit 😅

File: packages/product/src/repositories/product-tag.ts (hunk 1 above)
Comment (adrien2p) [reply packages/product/src/repositories/product-tag.ts]: the argument is already optional and have a default value being manager_

Comment (adrien2p) [reply packages/product/src/repositories/product-tag.ts]: wdyt of `@InjectEntityManager(doNotForceTransaction, false)`

**Comment by adrien2p** [comment]
> Do we need to type-cast all repositories going forward, because we use the DAL types on class members? It's fine with me, just messes up the code a bit 😅

Yes it is if you want to get the expected types since the interface does not know which type will be asked by the repository
but if you use the repository multiple times then you can just cast it once in a const and reuse it elsewhere

File: packages/product/src/repositories/product-tag.ts (hunk 1 above)
Comment (olivermrbl) [reply packages/product/src/repositories/product-tag.ts]: What are we telling the decorator when passing ""?

Comment (adrien2p) [reply packages/product/src/repositories/product-tag.ts]: that it should use `this.transaction` and not `this.something.transaction` so the string that is passed here represent the `something`

Comment (adrien2p) [reply packages/product/src/repositories/product-tag.ts]: and passing false would means that there is not a property to find on this and use this directly instead

Comment (adrien2p) [reply packages/product/src/repositories/product-tag.ts]: I would create a new parameter `useThis` if the argument was an object. but if I move to an object then I need to refactor all the existing modules. wdyt?

Comment (olivermrbl) [reply packages/product/src/repositories/product-tag.ts]: Let's keep it as is 👍

**Review by carlos-r-l-rodrigues — COMMENTED** [review]
Maybe if we use another property to be the manager on `base.ts` and asign `this.manager_ = this` that would force the InjectEntityManager find the default "manager_" property.

**Comment by adrien2p** [comment]
> Maybe if we use another property to be the manager on `base.ts` and assign `this.manager_ = this` that would force the InjectEntityManager find the default "manager_" property.

I thought about that but i came to the conclusion that it does not seems right since we are assigning the class to manager_ which seems weird and if a user want to use the decorator he will have to know that he can't use manager_ for its manager which seems strange

**Comment by adrien2p** [comment]
If the arguments was an object, we could have another option that force the default value to not be used?

**Comment by adrien2p** [comment]
@olivermrbl @carlos-r-l-rodrigues  I ve done the changes we talked about

**Pull request merged**